File: hindex.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        vector<int> dp(citations.size()+1, 0);
        for (auto &c: citations) {
            if (c >= citations.size()) dp.back()++;
            else dp[c]++;
        }
        int cur = 0;
        for (int i = dp.size()-1; i >= 0; i--) {
            cur += dp[i];
            if (cur >= i) return i;
        }
        return 0;
    }
};

File: python/hindex.py
class Solution(object):
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        l = len(citations)
        bucket = [0 for _ in xrange(l+1)]
        for citeNum in citations:
            if citeNum < l:
                bucket[citeNum] += 1
            else:
                bucket[-1] += 1
        t = 0
        res = 0
        for i in xrange(l,-1,-1):
            t += bucket[i]
            if t>=i:
                return i
        return 0
        
